packages/web-pkg/src/components/AppTopBar.vue

```diff
@@ -1,45 +1,151 @@
-    />
-    <div v-else />
-    <div>
-      <slot name="right" />
-      <oc-button
-        id="app-top-bar-close"
-        :aria-label="$gettext('Close')"
-        size="small"
-        @click="$emit('close')"
-      >
-        <oc-icon name="close" size="small" />
-      </oc-button>
+  <portal to="app.runtime.header.left">
+    <div class="oc-app-top-bar oc-flex">
+      <span class="oc-app-top-bar-inner oc-px-m oc-flex oc-flex-middle oc-flex-between">
+        <div class="open-file-bar oc-flex">
+          <oc-resource
+            v-if="resource?.name"
+            id="app-top-bar-resource"
+            :is-thumbnail-displayed="false"
+            :resource="resource"
+          />
+          <div v-else class="oc-resource oc-text-overflow">
+            <span v-text="$gettext('Loading')" />
+          </div>
+          <template v-if="dropDownActions.length">
+            <oc-button
+              id="oc-openfile-contextmenu-trigger"
+              v-oc-tooltip="contextMenuLabel"
+              :aria-label="contextMenuLabel"
+              appearance="raw"
+            >
+              <oc-icon name="more-2" />
+            </oc-button>
+            <oc-drop
+              drop-id="oc-openfile-contextmenu"
+              toggle="#oc-openfile-contextmenu-trigger"
+              mode="click"
+              close-on-click
+              padding-size="small"
+              @click.stop.prevent
+            >
+              <context-action-menu
+                :menu-sections="[{ name: 'dropdown-actions', items: dropDownActions }]"
+                :action-options="{ resources: [resource] }"
+              />
+            </oc-drop>
+          </template>
+        </div>
+
+        <div class="oc-flex main-actions">
+          <template v-if="mainActions.length">
+            <context-action-menu
+              :menu-sections="[{ name: 'main-actions', items: mainActions }]"
+              :action-options="{ resources: [resource] }"
+            />
+          </template>
+          <oc-button
+            id="app-top-bar-close"
+            v-oc-tooltip="closeButtonLabel"
+            class="oc-p-rm"
+            :aria-label="closeButtonLabel"
+            size="small"
+            appearance="raw"
+            @click="$emit('close')"
+          >
+            <oc-icon name="close" size="small" />
+          </oc-button>
+        </div>
+      </span>
     </div>
-  </div>
+  </portal>
 </template>
 
 <script lang="ts">
 import { defineComponent, PropType } from 'vue'
 import { Resource } from 'web-client/src'
+import { Action } from '../composables/actions/types'
+import ContextActionMenu from 'web-pkg/src/components/ContextActions/ContextActionMenu.vue'
 
 export default defineComponent({
   name: 'AppTopBar',
+  components: {
+    ContextActionMenu
+  },
   props: {
+    dropDownActions: {
+      type: Array as PropType<Action[]>,
+      default: () => []
+    },
+    mainActions: {
+      type: Array as PropType<Action[]>,
```

Comment:
should we limit this to a certain number of actions? if yes, then you could enforce that via a validator.

Reply:
I'm open for the suggestion, should I limit it to...one? Two?

Reply:
on second thought: it's the choice of the developer. If it looks awful or is unusable, they will reduce it themselves based on user feedback. So might be overengineered to limit the amount of main actions... and we don't need to document anywhere that it's limited to x main actions. :trollface: so let's keep it unlimited for now. sorry for the noise